src/site/stages/preview/index.js

```diff
@@ -0,0 +1,138 @@
+  //
+  // Unfortunately this must come before permalinks and navgation because of limitation in both
+  // modules regarding what files they understand. The consequence here is that liquid templates
+  // *within* a single file do NOT have access to the final path that they will be rendered under
+  // or any other metadata added by the permalinks() and navigation() filters.
+  //
+  // Thus far this has not been a problem because the only references to such paths are in the
+  // includes which are handled by the layout module. The layout module, luckily, can be run
+  // near the end of the filter chain and therefore has access to all the metadata.
+  //
+  // If this becomes a barrier in the future, permalinks should be patched to understand
+  // translating .md files which would allow inPlace() and markdown() to be moved under the
+  // permalinks() and navigation() filters making the variable stores uniform between inPlace()
+  // and layout().
+  smith.use(inPlace({ engine: 'liquid', pattern: '*.{md,html}' }));
+  smith.use(
+    markdown({
+      typographer: true,
+      html: true,
+    }),
+  );
+
+  // Responsible for create permalink structure. Most commonly used change foo.md to foo/index.html.
+  //
+  // This must come before navigation module, otherwise breadcrunmbs will see the wrong URLs.
+  //
+  // It also must come AFTER the markdown() module because it only recognizes .html files. See
+  // comment above the inPlace() module for explanation of effects on the metadata().
+  smith.use(
+    permalinks({
+      relative: false,
+      linksets: [
+        {
+          match: { collection: 'posts' },
+          pattern: ':date/:slug',
+        },
+      ],
+    }),
+  );
+
+  smith.use(
+    navigation({
+      navConfigs: {
+        sortByNameFirst: true,
+        breadcrumbProperty: 'breadcrumb_path',
+        pathProperty: 'nav_path',
+        includeDirs: true,
+      },
+      navSettings: {},
+    }),
+  );
+
+  smith.use(
+    layouts({
+      engine: 'liquid',
+      directory: BUILD_OPTIONS.layouts,
+      // Only apply layouts to markdown and html files.
+      pattern: '**/*.{md,html}',
+    }),
+  );
+
+  /*
+  Add nonce attribute with substition string to all inline script tags
+  Convert onclick event handles into nonced script tags
+  */
+  smith.use(nonceTransformer);
+
+  /*
+  * This will replace links in static pages with a staging domain,
+  * if it is in the list of domains to replace
+  */
+  smith.use(rewriteVaDomains(BUILD_OPTIONS));
+
+  // Create the data passed from the content build to the assets compiler.
+  // On the server, it can be accessed at BUILD_OPTIONS.buildSettings.
+  // In the browser, it can be accessed at window.settings.
+  smith.use(createBuildSettings(BUILD_OPTIONS));
+
+  smith.use(updateExternalLinks(BUILD_OPTIONS));
+
+  // smith.use(addAssetHashes(BUILD_OPTIONS));
+
+  return smith;
+}
+
+module.exports = createPipeline;
```

Comment:
Is it possible to share this pipeline/function with `src/stages/build/index.js`? At a glance it looks like the pipelines are pretty similar.

Reply:
I looked into that a little bit and it would take some reorganization to make that work. There's not one single block that is the same as the build pipeline. I definitely want to make them share as much code as possible, but it felt like it was enough work to be a separate ticket.